### PR TITLE
Search: Move entity events to transaction

### LINKED ISSUE #52212: Panel Search not working for imported Dashboard via API
**What happened**: Panel Search is not working correctly for API imported dashboard. As soon as you search panels shows up, but instead of linking to the full view of the panel it redirect you to the dashboard containing the panel. As soon as you edit and save the dashboard the search start working correctly.

**How to reproduce it**: Import one dashboard via api, search the panel inside the imported dashboard. Search result has wrong link, all pointing too ?viewPanel=0, as soon you edit the dashboard it start working.

- Grafana version: 9.0.2

### PR DESCRIPTION
**Pull request by FZambia**
**What this PR does / why we need it**:

Previously we were emitting entity events in API handlers, mostly because we wanted to be non-obtrusive while we are evaluating the new search approach.

To reduce a chance of missing events and have a broken state till the next re-index we are moving event emission to the transaction. This also helps to avoid problems with forgotten event emission in some API that deals with dashboards and folders.

This covers dashboards save/delete and folders save/delete.

Fixes #52212 

This pr also removes search event local broadcasting logic. It did not work for HA case – so at this moment we are using https://github.com/grafana/grafana/pull/50152 approach.

### CODE REVIEW

File: pkg/services/dashboards/database/database.go [hunk 1]
```diff
@@ -812,6 +833,25 @@ func (d *DashboardStore) deleteDashboard(cmd *models.DeleteDashboardCommand, ses
...
+			Created:   time.Now().Unix(),
+		}
+	} else {
+		entityEvent = &store.EntityEvent{
+			EventType: store.EntityEventTypeDelete,
+			EntityId:  store.CreateDatabaseEntityId(dashboard.Uid, dashboard.OrgId, store.EntityTypeDashboard),
+			Created:   time.Now().Unix(),
+		}
+	}
+	_, err = sess.Insert(entityEvent)
```
[the remark this answers is not in the document]
Comment (FZambia) [reply pkg/services/dashboards/database/database.go:850]: Automatically by a caller code when error is returned

File: pkg/services/dashboards/database/database.go [hunk 2]
```diff
@@ -505,6 +507,25 @@ func saveDashboard(sess *sqlstore.DBSession, cmd *models.SaveDashboardCommand) e
...
+			Created:   time.Now().Unix(),
+		}
+	} else {
+		entityEvent = &store.EntityEvent{
+			EventType: store.EntityEventTypeUpdate,
+			EntityId:  store.CreateDatabaseEntityId(dash.Uid, dash.OrgId, store.EntityTypeDashboard),
+			Created:   time.Now().Unix(),
+		}
+	}
+	_, err = sess.Insert(entityEvent)
```
Comment (ArturWierzbicki) [inline pkg/services/dashboards/database/database.go:524]: nit: could be extracted to thats used in both places

```suggestion
	_, err = sess.Insert(createEntityEvent(dash, store.EntityEventTypeUpdate))
```

File: pkg/services/store/entity_events.go [hunk 3]
```diff
@@ -143,6 +111,18 @@ func (e *entityEventService) GetAllEventsAfter(ctx context.Context, id int64) ([
 	return evs, err
 }
 
+func (e *entityEventService) saveEvent(ctx context.Context, cmd SaveEventCmd) error {
```
Comment (ArturWierzbicki) [inline pkg/services/store/entity_events.go:114]: hmm.. this method is now used only for tests - should we just remove it and have a different setup in tests?

**Review by ArturWierzbicki — APPROVED** [review]
have tested just the few basic scenarios (dashboard/folder crud via UI) but it looks great! 🚀

**Review by ryantxu — APPROVED** [review]

**Comment by FZambia** [comment]
Forgot to check feature toggle!

File: pkg/services/dashboards/database/database.go (hunk 2 above)
Comment (FZambia) [reply pkg/services/dashboards/database/database.go]: Done in 78c7b881e1ae857d65376782f24941a30c9e6930

File: pkg/services/store/entity_events.go [hunk 4]
```diff
@@ -90,36 +88,6 @@ type entityEventService struct {
...
-		_, err := sess.Insert(entityEvent)
-		return err
-	})
-	if err != nil {
-		return err
-	}
-	return e.broadcastEvent(ctx, entityEvent)
-}
-
-func (e *entityEventService) broadcastEvent(ctx context.Context, event *EntityEvent) error {
```
Comment (ying-jeanne) [inline pkg/services/store/entity_events.go:109]: I am not following completely the new search. The code you removed actually do insert event into entity_event table, and then broadcast event after. In the new added code, there is no longer broadcast event?

Comment (FZambia) [reply pkg/services/store/entity_events.go:109]: Hello, yep - we are not broadcasting it anymore. This was a way to achieve consistent search reads after changing dashboard. But it did not work for HA case. So at this moment we switched to https://github.com/grafana/grafana/pull/50152 approach.

I'll update PR description - had to mention that this pr also remove broadcasting logic.

File: pkg/services/store/entity_events.go (hunk 3 above)
Comment (FZambia) [reply pkg/services/store/entity_events.go]: Done in f8f6a24ddcce9e54979314b2bd2e22bb00d109d8

**Comment by ying-jeanne** [comment]
Talked with @zserge today, I saw that in the PR dashboard service is going to do insert into entity event table, it is not really what backend is trying to do right now, since we want one table only be able to be read/written by one service. But we haven't decide how to manage the transaction between different services yet, and in the future the event probably would be managed in another way than insert into sql table, the implementation seems reasonable.

**Review by ying-jeanne — APPROVED** [review]

**Pull request merged**